src/Material/Menu.elm

```diff
@@ -0,0 +1,523 @@
+-}
+
+import Dict exposing (Dict)
+import Effects exposing (Effects, none)
+import Html.Attributes as Html exposing (..)
+import Html.Events as Html exposing (defaultOptions)
+import Html exposing (..)
+import Json.Decode as Json exposing (Decoder)
+import Json.Encode exposing (string)
+import Material.Helpers exposing (..)
+import String
+import Task
+
+import Material.Menu.Oracle as Oracle exposing (Oracle)
+import Material.Ripple as Ripple
+import Material.Style as Style exposing (Style, cs, cs', css, css', styled)
+
+{-| MDL menu.
+-}
+
+
+-- CONSTANTS
+
+constant :
+  { transitionDurationSeconds  : Float
+  , transitionDurationFraction : Float
+  , closeTimeout               : Float
+  }
+constant =
+  { transitionDurationSeconds  = 0.3
+  , transitionDurationFraction = 0.8
+  , closeTimeout               = 150
+  }
+
+
+-- TODO: Key codes are not implemented yet.
+
+--keycodes =
+--  { enter     = 13
+--  , escape    = 27
+--  , space     = 32
+--  , upArrow   = 38
+--  , downArrow = 40
+--  }
+
+
+-- MODEL
+
+
+{-| Model of the menu; common to all kinds of menus.
+Use `model` to initialise it.
+-}
+
+type alias Model =
+  { alignment : Style
+  , ripple : Bool
+  , items : Dict Int Ripple.Model
+  , open : Maybe Bool --
+  , closing : Bool
+  , oracle : Maybe Oracle
+  }
+
+
+{-| Model initialiser. Call with `True` if the menu items should ripple when clicked, `False` otherwise. Pass one of `bottomLeft`, `bottomRight`, `topLeft`, `topRight` to specify the menu's alignment in relation to the button. Use `unaligned` if you want to align it manually.
+-}
+
+model : Bool -> Style -> Model
+model ripple alignemnt =
+  { alignment = alignemnt
+  , ripple = ripple
+  , items = Dict.empty
+  , open = Nothing
+  , closing = False
+  , oracle = Nothing
+  }
+
+
+{-| Component model styles.
+Used with `model`.
+-}
+
+bottomLeft : Style
+bottomLeft = cs "mdl-menu--bottom-left"
+
+bottomRight : Style
+bottomRight = cs "mdl-menu--bottom-right"
+
+topLeft : Style
+topLeft = cs "mdl-menu--top-left"
+
+topRight : Style
+topRight = cs "mdl-menu--top-right"
+
+unaligned : Style
+unaligned = cs "mdl-menu--unaligned"
+
+
+-- ITEM
+
+
+{-| Item model.
+-}
+
+type alias Item =
+  { html    : Html
+  , enabled : Bool
+  , divider : Bool
+  }
+
+{-| Item constructor. Pass `True` if there should be a divider below this item; `False` otherwise. Pass as second argument `True` if item should be enabled, and pass as third argument any `Html` as content of the menu item.
+-}
+
+item : Bool -> Bool -> Html -> Item
+item divider enabled html =
+  { html = html
+  , enabled = enabled
+  , divider = divider
+  }
+
+
+-- ACTION, UPDATE
+
+
+{-| Component action. The `Select n` action fires when the `n`th Item is selected; `Open`, `Close` when the menu opens, closes.
+-}
+
+-- Note: The Close' index is 1-based while the Select index is 0-based.
+
+type Action =
+    Toggle Oracle
+  | Close' Int Oracle
+  | Hide Int Oracle
+
+  | Select Int
```

Comment:
Would it be an idea to have `Item` contain the action to dispatch when the item is selected, rather than giving the `Select n` action? I'm not sure; I did something like that in [Snackbar](https://github.com/debois/elm-mdl/blob/master/src/Material/Snackbar.elm#L49-L71), but I'm not sure I liked how that turned out. 

(It makes it difficult to work with component support.)